eth/borfinality/whitelist/milestone.go

```diff
@@ -0,0 +1,312 @@
+// 	//Metrics for collecting the whitelisted milestone number
+// 	whitelistedMilestoneMeter = metrics.NewRegisteredGauge("chain/milestone/latest", nil)
+
+// 	//Metrics for collecting the future milestone number
+// 	FutureMilestoneMeter = metrics.NewRegisteredGauge("chain/milestone/future", nil)
+
+// 	//Metrics for collecting the length of the MilestoneIds map
+// 	MilestoneIdsLengthMeter = metrics.NewRegisteredGauge("chain/milestone/idslength", nil)
+
+// 	//Metrics for collecting the number of valid chains received
+// 	MilestoneChainMeter = metrics.NewRegisteredMeter("chain/milestone/isvalidchain", nil)
+
+// 	//Metrics for collecting the number of valid peers received
+// 	MilestonePeerMeter = metrics.NewRegisteredMeter("chain/milestone/isvalidpeer", nil)
+// )
+
+// IsValidChain checks the validity of chain by comparing it
+// against the local milestone entries
+func (m *milestone) IsValidChain(currentHeader *types.Header, chain []*types.Header) (bool, error) {
+	//Checking for the milestone flag
+	if !flags.Milestone {
+		return true, nil
+	}
+
+	m.finality.RLock()
+	defer m.finality.RUnlock()
+
+	// TODO: Uncomment once metrics is added
+	// var isValid bool = false
+	// defer func() {
+	// 	if isValid {
+	// 		MilestoneChainMeter.Mark(int64(1))
+	// 	} else {
+	// 		MilestoneChainMeter.Mark(int64(-1))
+	// 	}
+	// }()
+
+	res, err := m.finality.IsValidChain(currentHeader, chain)
+
+	if !res {
+		// isValid = false
+		return false, err
+	}
+
+	if m.Locked && !m.IsReorgAllowed(chain, m.LockedMilestoneNumber, m.LockedMilestoneHash) {
+		// isValid = false
+		return false, nil
+	}
+
+	if !m.IsFutureMilestoneCompatible(chain) {
+		// isValid = false
+		return false, nil
+	}
+
+	// isValid = true
+
+	return true, nil
+}
+
+// IsValidPeer checks if the chain we're about to receive from a peer is valid or not
+// in terms of reorgs. We won't reorg beyond the last bor finality submitted to mainchain.
+func (m *milestone) IsValidPeer(fetchHeadersByNumber func(number uint64, amount int, skip int, reverse bool) ([]*types.Header, []common.Hash, error)) (bool, error) {
+	if !flags.Milestone {
+		return true, nil
+	}
+
+	res, err := m.finality.IsValidPeer(fetchHeadersByNumber)
+
+	// TODO: Uncomment once metrics is added
+	// if res {
+	// 	MilestonePeerMeter.Mark(int64(1))
+	// } else {
+	// 	MilestonePeerMeter.Mark(int64(-1))
+	// }
+
+	return res, err
+}
+
+func (m *milestone) Process(block uint64, hash common.Hash) {
+	m.finality.Lock()
+	defer m.finality.Unlock()
+
+	m.finality.Process(block, hash)
+
+	for i := 0; i < len(m.FutureMilestoneOrder); i++ {
+		if m.FutureMilestoneOrder[i] <= block {
+			m.dequeueFutureMilestone()
+		} else {
+			break
+		}
+	}
+
+	// TODO: Uncomment once metrics is added
+	// whitelistedMilestoneMeter.Update(int64(block))
+
+	m.UnlockSprint(block)
+}
+
+// This function will Lock the mutex at the time of voting
+// fixme: get rid of it
+func (m *milestone) LockMutex(endBlockNum uint64) bool {
+	m.finality.Lock()
```

Comment:
Can you confirm once if m.finality.Unlock() is not needed?

Reply:
We have a separate function `UnlockMutex`